### PR TITLE
nixos/restic: generalize cache configuration

### PR DESCRIPTION
**Pull request by oxzi**
###### Description of changes

The restic repository cache location defaults to ~/.cache/restic when not overwritten either by the --cache-dir command line parameter or the universal RESTIC_CACHE_DIR environment variable.

Currently, the --cache-dir variable is set to only some restic commands, but, e.g., not to the unit's preStart command for the module's initialize option. This results in two distinct cache locations, one at ~/.cache/restic for the initialize commands and one at the configured --cache-dir location for the restic backup command.

By explicitly setting RESTIC_CACHE_DIR for the unit, only one cache at the correct location will be used.

https://restic.readthedocs.io/en/v0.15.1/manual_rest.html#caching

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/restic.nix [hunk 1]
```diff
@@ -129,8 +129,8 @@ import ./make-test-python.nix (
 
           # test that custompackage runs both `restic backup` and `restic check` with reasonable commandlines
           "systemctl start restic-backups-custompackage.service",
-          "grep 'backup .* /opt' /tmp/fake-restic.log",
-          "grep 'check .* --some-check-option' /tmp/fake-restic.log",
+          "grep 'backup.* /opt' /tmp/fake-restic.log",
+          "grep 'check.* --some-check-option' /tmp/fake-restic.log",
```
Comment (Lassulus) [inline nixos/tests/restic.nix:133]: why was this changed?

Comment (oxzi) [reply nixos/tests/restic.nix:133]: Thanks for your feedback!

The old search string ` .* ` had spaces before and after an optional command line argument. However, with having spaces on both sides the shortest accepted string would be `  ` (two spaces) or, in other words, at least one argument was required.
Prior to this PR's changes, most restic commands had the `--cache-dir=%C/restic-backups-${name}` command line argument, which would satisfy the `grep` command in question. After moving the cache location to the environment variable, present now to all restic calls of this module and not only most of them, the default `restic backup` call now has no further command line arguments and also `restic check` has lost one.
